packages/contract/src/contracts/functions/base-invocation-scope.ts

```diff
@@ -0,0 +1,214 @@
+        amount: toBigInt(call.amount || 0),
+      }))
+      // Add required amount to pay gas fee
+      .concat({
+        assetId: NativeAssetId,
+        amount: 1n,
+      })
+      .filter(({ assetId, amount }) => assetId && amount);
+    return assets;
+  }
+
+  protected updateRequiredCoins() {
+    const assets = this.getRequiredCoins();
+    const reduceForwardCoins = (
+      requiredCoins: Map<any, CoinQuantity>,
+      { assetId, amount }: CoinQuantity
+    ) => {
+      const currentAmount = requiredCoins.get(assetId)?.amount || 0n;
+
+      return requiredCoins.set(assetId, {
+        assetId: String(assetId),
+        amount: currentAmount + toBigInt(amount),
+      });
+    };
+    this.requiredCoins = Array.from(
+      assets.reduce(reduceForwardCoins, new Map<any, CoinQuantity>()).values()
+    );
+  }
+
+  protected addCall(funcScope: InvocationScopeLike) {
+    this.addCalls([funcScope]);
+    return this;
+  }
+
+  protected addCalls(funcScopes: Array<InvocationScopeLike>) {
+    this.functionInvocationScopes.push(...funcScopes);
+    this.updateScriptRequest();
+    this.updateRequiredCoins();
+    return this;
+  }
+
+  protected async prepareTransaction(options?: CallOptions) {
+    // Update request scripts before call
+    this.updateScriptRequest();
+
+    // Add funds required on forwards and to pay gas
+    const opts = BaseInvocationScope.getCallOptions(options);
+    if (opts.fundTransaction && this.contract.wallet) {
+      await this.fundWithRequiredCoins();
+    }
+  }
+
+  /**
+   * Add to the transaction scope the required amount of unspent UTXO's.
+   *
+   * Required Amount = forward coins + transfers + gas fee.
+   */
+  async fundWithRequiredCoins() {
+    // Clean coin inputs before add new coins to the request
+    this.transactionRequest.inputs = this.transactionRequest.inputs.filter(
+      (i) => i.type !== InputType.Coin
+    );
+    const coins = await this.contract.wallet?.getCoinsToSpend(this.requiredCoins);
+    this.transactionRequest.addCoins(coins || []);
+    return this;
+  }
+
+  txParams(txParams: TxParams) {
+    this.txParameters = txParams;
+    const request = this.transactionRequest;
+
+    request.gasLimit = toBigInt(txParams.gasLimit || request.gasLimit);
+    request.gasPrice = toBigInt(txParams.gasPrice || request.gasPrice);
+    request.bytePrice = toBigInt(txParams.bytePrice || request.bytePrice);
+    request.addVariableOutputs(this.txParameters?.variableOutputs || 0);
+
+    return this;
+  }
+
+  /**
+   * Submits a transaction to the blockchain.
+   *
+   * This is a final action and will spend the coins and change the state of the contract.
+   * It also means that invalid transactions will throw an error, and consume gas. To avoid this
+   * running invalid tx and consuming gas try to `simulate` first when possible.
+   */
+  async call<T = TReturn>(options?: CallOptions): Promise<FunctionInvocationResult<T>> {
+    assert(this.contract.wallet, 'Wallet is required!');
```

Comment:
This naming here is odd. Assertions are used to check invariants, and are preferably removed from production code using something like [babel-plugin-unassert](https://github.com/unassert-js/babel-plugin-unassert).

An assertion would make sense here if, for example, it was impossible for a user to call `call` without a `this.contract.wallet`, but TypeScript didn't understand this so we still had to check to make TS happy. We don't have such a case here since it's very possible for a user to create a wallet-less Contract and call `call`. We should just do regular error handling here.

https://en.wikipedia.org/wiki/Assertion_(software_development)#Comparison_with_error_handling

Reply:
This assertion is a custom method from `./utils` but maybe is a good idea to change the name.

Reply:
Also on the other point, it's not possible to `submit a valid tx` without a wallet instance, as we need it to add funds and pay the gas fee. The only time is possible to submit a tx without a wallet is with `dryRun` in this case we have other methods such as `dryRun` and `get`.

Reply:
By possibility I specifically mean whether it's possible for the assertion in question to fail.

For example:
```ts
  async call<T = TReturn>(options?: CallOptions): Promise<FunctionInvocationResult<T>> {
    // We do regular error handling here because the user can create a Contract
    // without a wallet and call `call`.
    // In production, we can't remove this check. We need to properly notify the user
    // about what was wrong.
    if (!this.contract.wallet) {
      throw new Error('Wallet is required!');
    }
    
    return this.#call(options);
  }

  async #call<T = TReturn>(options?: CallOptions): Promise<FunctionInvocationResult<T>> {
    // We do an assertion here because even though TypeScript thinks `this.contract.wallet`
    // can be undefined, it is impossible for this to happen.
    // We call this method from only one place and we make this check already.
    // In production, we can safely remove this check.
    assert(this.contract.wallet, 'Expected a wallet.');

    await this.prepareTransaction(options);
    const response = await this.contract.wallet.sendTransaction(this.transactionRequest);

    return FunctionInvocationResult.build<T>(
      this.functionInvocationScopes,
      response,
      this.isMultiCall
    );
  }
  ```